[#117] Warn on duplicate properties

ghillairet opened the pull request:
See #117

ghillairet committed:
[#117] Warn on duplicate properties

ghillairet committed:
[#117] Warn on duplicate properties

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java:
```diff
@@ -56,21 +66,85 @@
+						duplicates.get(key).add(keyNode);
+					} else {
+						duplicates.put(key, Sets.newHashSet(keyNode));
+					}
```
I don't know why you don't like Guava, but sometimes it's very useful. A `Multimap` would save us a four lines of the if-else here:

``` java
Multimap<String, Node> duplicates = HashMultimap.<String, Node>create();
<...>
duplicates.put(key, keyNode);
```

tfesenko commented on com.reprezen.swagedit.tests/src/com/reprezen/swagedit/tests/ValidatorTest.xtend:
```diff
@@ -187,4 +189,89 @@ class ValidatorTest {
+			new SwaggerError(10, IMarker.SEVERITY_WARNING, String.format(Messages.error_duplicate_keys, "in")),
+			new SwaggerError(11, IMarker.SEVERITY_WARNING, String.format(Messages.error_duplicate_keys, "in"))
+		))
+	}
```
I would also add a test for a case when the value is an object, for example here `responses` defined twice within the same pathItem:

``` yaml
swagger: '2.0'
info:
  version: 0.0.0
  title: Simple API
paths:
  /foo/{bar}:
    get:
      responses:
        '200':
          description: OK
      responses:
        '201':
          description: OK
```

It works as expected - validator shows a warning, but we don't have a test for it.

tfesenko replied:
Thanks for the tests!

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java:
```diff
@@ -56,21 +66,85 @@
+								IMarker.SEVERITY_WARNING, 
+								String.format(Messages.error_duplicate_keys, key)));
+					}
+				}
```
I am thinking of a burden on performance. We have two passes here:
1. to create a map
2. to check that every key has a single value.
@ghillairet , let's think how we can improve it. Maybe, use a map that would throw an exception if you add a value for the same key twice and catch an exception.

ghillairet replied:
If you throw the exception on the first key encounter twice, how do you handle cases where an object has multiple different duplicate keys?
What I can do is collect all duplicates in a single map and resolve duplicates once after the document is parsed.

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java:
```diff
@@ -56,21 +66,85 @@
+			for (NodeTuple tuple: n.getValue()) {
+				Node keyNode = tuple.getKeyNode();
+
+				if (keyNode.getNodeId() == NodeId.scalar) {
```
I wonder why we need a check for `scalar` here

ghillairet committed:
[#117] Warn on duplicate properties

Rewrite the implementation of the duplicate keys validation method to avoid excessive iteration. The method now uses a multi map of pair<Node, String> x Set<Node>. With the pair left element being a parent node and the pair right element being a key value, the Set of node contains all children of the parent node being a key node and whose value matches the pair right element. Resolution of duplicates is done only once after all of them are collected in the map.

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java:
```diff
@@ -101,50 +97,65 @@
+	 * the pair's key.
+	 * Once the iteration is done, the resulting map should be traversed. Each pair having 
+	 * more than one element in its associated Set are duplicate keys. 
+	 */
```
This is a comment, not a Java doc. I wonder if it was your intention.
Similar with other methods in this class.
